google/cloud/bigtable/admin/wait_for_consistency_test.cc

```diff
@@ -0,0 +1,381 @@
+
+  internal::AutomaticallyCreatedBackgroundThreads background;
+  auto mock = std::make_shared<MockConnection>();
+  auto client = BigtableTableAdminClient(mock);
+
+  EXPECT_CALL(*mock, AsyncCheckConsistency)
+      .Times(kLimitedErrorCount + 1)
+      .WillRepeatedly([&](btadmin::CheckConsistencyRequest const& request) {
+        EXPECT_EQ(request.name(), table_name);
+        EXPECT_EQ(request.consistency_token(), token);
+        return make_ready_future<StatusOr<btadmin::CheckConsistencyResponse>>(
+            Status(StatusCode::kUnavailable, "try again"));
+      });
+
+  auto status = AsyncWaitForConsistency(background.cq(), client, table_name,
+                                        token, TestOptions())
+                    .get();
+  EXPECT_THAT(status,
+              StatusIs(StatusCode::kUnavailable, HasSubstr("try again")));
+}
+
+TEST(AsyncWaitForConsistency, NeverConsistent) {
+  std::string const table_name =
+      bigtable::TableName("test-project", "test-instance", "test-table");
+  std::string const token = "test-token";
+
+  internal::AutomaticallyCreatedBackgroundThreads background;
+  auto mock = std::make_shared<MockConnection>();
+  auto client = BigtableTableAdminClient(mock);
+
+  EXPECT_CALL(*mock, AsyncCheckConsistency)
+      .Times(kLimitedErrorCount + 1)
+      .WillRepeatedly([&](btadmin::CheckConsistencyRequest const& request) {
+        EXPECT_EQ(request.name(), table_name);
+        EXPECT_EQ(request.consistency_token(), token);
+        btadmin::CheckConsistencyResponse response;
+        response.set_consistent(false);
+        return make_ready_future(make_status_or(response));
+      });
+
+  auto status = AsyncWaitForConsistency(background.cq(), client, table_name,
+                                        token, TestOptions())
+                    .get();
+  EXPECT_THAT(status, StatusIs(StatusCode::kDeadlineExceeded,
+                               HasSubstr("Polling loop terminated")));
+}
+
+class AsyncWaitForConsistencyCancelTest : public ::testing::Test {
+ protected:
+  int CancelCount() {
+    std::lock_guard<std::mutex> lk(mu_);
+    int n = cancel_count_;
+    cancel_count_ = 0;
+    return n;
+  }
+
+  future<StatusOr<btadmin::CheckConsistencyResponse>> SimulateRequest() {
+    promise<bool> p([this] {
+      std::lock_guard<std::mutex> lk(mu_);
+      ++cancel_count_;
+    });
+    auto f = p.get_future().then([](future<bool> g) {
+      btadmin::CheckConsistencyResponse r;
+      r.set_consistent(g.get());
+      return make_status_or(r);
+    });
+    std::lock_guard<std::mutex> lk(mu_);
+    requests_.push_back(std::move(p));
+    cv_.notify_one();
+    return f;
+  }
+
+  promise<bool> WaitForRequest() {
+    std::unique_lock<std::mutex> lk(mu_);
+    cv_.wait(lk, [&] { return !requests_.empty(); });
+    auto p = std::move(requests_.front());
+    requests_.pop_front();
+    return p;
+  }
+
+ private:
+  // We will simulate the asynchronous requests by pushing promises into this
```

Comment:
There is a helper class to do this sort of thing:

https://github.com/googleapis/google-cloud-cpp/blob/5d19d3d16eb458414ff96162f3080b09d40369df/google/cloud/testing_util/async_sequencer.h#L30-L65

If it does not work in this case, maybe add a comment explaining why?

Reply:
Ah, thanks for the reminder. I opened #7771 to add the cancel functionality to `AsyncSequencer`.

I propose that I leave the PR as is for now. The code in this test is stolen from [`async_retry_loop_test.cc`](https://github.com/googleapis/google-cloud-cpp/blob/5d19d3d16eb458414ff96162f3080b09d40369df/google/cloud/internal/async_retry_loop_test.cc#L58-L98).  I think it will be easiest to cleanup both of these test fixtures at the same time. (FYI: cleanup means using `AsyncSequencer` and `MockCompletionQueueImpl`)

Reply:
SGTM, as long as we have an issue for this cleanup.